### PR TITLE
Introduce a new error framework

### PR DESCRIPTION
**Pull request by breeswish**
Also changed some part of deploy command to use this framework:

image
image

note:

- Please switch to use `zap.L()` for audit and debug logging, which separates logs and console display. `log.Output` is thus removed.
- `tiops template` is not yet implemented (TODO).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,3 +1,5 @@
+tiops-debug-*.log
```
Comment (lonng) [inline .gitignore:1]: We should use the name `tiup-cluster-debug-*.log`

File: cmd/deploy.go [hunk 2]
```diff
@@ -53,27 +63,47 @@ func newDeploy() *cobra.Command {
...
 			return deploy(args[0], args[1], args[2], opt)
 		},
 	}
 
 	cmd.Flags().StringVar(&opt.user, "user", "root", "Specify the system user name")
 	cmd.Flags().BoolVar(&opt.usePasswd, "password", false, "Specify the password of system user")
-	cmd.Flags().StringVar(&opt.keyFile, "key", "", "Specify the key path of system user")
-	cmd.Flags().StringVar(&opt.passphrase, "passphrase", "", "Specify the passphrase of the key")
-	cmd.Flags().BoolVar(&opt.skipConfirm, "noconfirm", false, "Skip the confirmation of topology")
+	cmd.Flags().StringVar(&opt.keyFile, "identity_file", "", "Specify the path of the SSH identity file")
```
Comment (lonng) [inline cmd/deploy.go:103]:
```suggestion
	cmd.Flags().StringVarP(&opt.keyFile, "identity_file", "i", "", "Specify the path of the SSH identity file")
```

File: cmd/root.go [hunk 3]
```diff
@@ -94,33 +89,105 @@ func init() {
 	)
 }
 
+func printErrorMessageForNormalError(err error) {
+	_, _ = colorutil.ColorErrorMsg.Fprintf(os.Stderr, "\nError: %s\n", err.Error())
```
Comment (lonng) [inline cmd/root.go:93]: I think it's better to provide a debug mode to output the backtrace. The former implementation has a `TIUP_BACKTRACE` to enable backtrace.

File: cmd/root.go [hunk 4]
```diff
@@ -94,33 +89,105 @@ func init() {
...
+		} else {
+			break
+		}
+	}
+	_, _ = colorutil.ColorErrorMsg.Fprintf(os.Stderr, "\nError: %s", msg)
+}
+
 // Execute executes the root command
 func Execute() {
+	zap.L().Info("Execute command", zap.Strings("args", os.Args))
```
Comment (lonng) [inline cmd/root.go:126]: We should insert a `tiup` in front of `os.Args` if we run tiops in the tiup mode. e.g: `tiup cluster deploy ...`.
Just need to move this line to L134 and we can determine whether tiops run as a tiup component via `wd != ""`.

File: cmd/scale_out.go [hunk 5]
```diff
@@ -14,19 +14,21 @@
...
 import (
+	"fmt"
 	"path/filepath"
 	"strings"
 
+	"github.com/pingcap-incubator/tiup/pkg/set"
+	tiuputils "github.com/pingcap-incubator/tiup/pkg/utils"
+	"github.com/pingcap/errors"
+	"github.com/spf13/cobra"
+
```
Comment (lonng) [inline cmd/scale_out.go:25]: Remove this blank line.

File: cmd/upgrade.go [hunk 6]
```diff
@@ -18,15 +18,17 @@ import (
...
-	"github.com/pingcap-incubator/tiops/pkg/bindversion"
-	"github.com/pingcap-incubator/tiops/pkg/meta"
-	operator "github.com/pingcap-incubator/tiops/pkg/operation"
-	"github.com/pingcap-incubator/tiops/pkg/task"
 	"github.com/pingcap-incubator/tiup/pkg/repository"
 	"github.com/pingcap-incubator/tiup/pkg/utils"
 	"github.com/pingcap/errors"
 	"github.com/spf13/cobra"
 	"golang.org/x/mod/semver"
+
```
Comment (lonng) [inline cmd/upgrade.go:26]: ditto

File: pkg/cliutil/cliutil.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"strings"
+	"text/template"
+
+	"github.com/joomcode/errorx"
+	"github.com/spf13/cobra"
+
```
Comment (lonng) [inline pkg/cliutil/cliutil.go:25]: ditto

File: pkg/logger/audit.go [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+	"bytes"
+	"io/ioutil"
+	"os"
+	"strings"
+	"time"
+
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
```
Comment (lonng) [inline pkg/logger/audit.go:26]: ditto

File: pkg/logger/debug.go [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"time"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
```
Comment (lonng) [inline pkg/logger/debug.go:26]: ditto

File: pkg/logger/debug.go [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+func newDebugLogCore() zapcore.Core {
+	debugBuffer = new(bytes.Buffer)
+	encoder := zapcore.NewConsoleEncoder(zap.NewDevelopmentEncoderConfig())
+	return zapcore.NewCore(encoder, zapcore.Lock(zapcore.AddSync(debugBuffer)), zapcore.DebugLevel)
+}
+
+// OutputDebugLog outputs debug log in the current working directory.
+func OutputDebugLog() {
+	// FIXME: Stupid go does not allow writing fraction seconds without a leading dot.
+	fileName := time.Now().Format("./tiops-debug-2006-01-02-15-04-05.log")
```
Comment (lonng) [inline pkg/logger/debug.go:41]:
```suggestion
	fileName := time.Now().Format("./tiiup-cluster-debug-2006-01-02-15-04-05.log")
```

File: pkg/utils/topology.go [hunk 11]
```diff
@@ -2,19 +2,67 @@ package utils
...
 import (
 	"io/ioutil"
+	"strings"
 
-	"github.com/pingcap/errors"
-	"gopkg.in/yaml.v2"
+	"github.com/goccy/go-yaml"
+	"github.com/joomcode/errorx"
+	"go.uber.org/zap"
+
```
Comment (lonng) [inline pkg/utils/topology.go:10]: ditto

File: cmd/root.go (hunk 3 above)
Comment (breeswish) [reply cmd/root.go:93]: In debug logs there will be backtraces.

**Pull request merged**